harness/determined/experimental/core_v2.py

```diff
@@ -0,0 +1,229 @@
+    # - min_validation_period
+    # - pbs
+    # - perform_initial_validation
+    # - records_per_epoch
+    # - scheduling_unit
+    # Deprecated:
+    # - data_layer
+    # - pbs
+    # - tensorboard_storage
+    checkpoint_storage: Optional[Union[str, Dict[str, Any]]] = None
+    # Searcher is currently a hack to disambiguate single trial experiments
+    # and hp searches in WebUI.
+    # Also searcher metric is useful for sorting in the UI and, in the future, checkpoint gc.
+    searcher: Optional[Dict[str, Any]] = None
+    # TODO(ilia): later to be replaced with:
+    # Unmanaged mode only config options:
+    # multi_trial_experiment: bool = False
+    # metric: Optional[str] = None
+    # smaller_is_better: bool = True  # mode?
+
+
+@dataclasses.dataclass
+class UnmanagedConfig:
+    """
+    `UnmanagedConfig` values are only used in the unmanaged mode.
+    """
+
+    # For the managed mode, workspace is critical for RBAC so it cannot be easily
+    # merged and patched during the experiment runtime.
+    workspace: Optional[str] = None
+    project: Optional[str] = None
+    # External experiment & trial ids.
+    # `external_experiment_id` is used to uniquely identify an experiment when grouping
+    # multiple trials as one HP search, or if any trial within this experiment will be resumed.
+    # `external_trial_id` is used to uniquely identify the resumed trial within the experiment.
+    # if `external_trial_id` is specified, `external_experiment_id` MUST be passed as well.
+    #
+    # If you are going to resume trials, whether in hp search or single-trial experiments,
+    # specify both `external_experiment_id` and `external_trial_id`.
+    # If you are going to use hp search, but will not resume trials,
+    # specifing the `external_experiment_id` is sufficient.
+    # If you are not going to use either feature, omit these options.
+    external_experiment_id: Optional[str] = None
+    external_trial_id: Optional[str] = None
+
+
+def _set_globals() -> None:
+    global train
+    global checkpoint
+    global distributed
+    global preempt
+    global searcher
+    global info
+
+    assert _context is not None
+    train = _context.train
+    checkpoint = _context.checkpoint
+    distributed = _context.distributed
+    preempt = _context.preempt
+    searcher = _context.searcher
+    info = _context.info
+
+
+def init(
+    *,
+    defaults: Optional[DefaultConfig] = None,
+    unmanaged: Optional[UnmanagedConfig] = None,
+    client: Optional[Determined] = None,
+    # Classic core context arguments.
+    distributed: Optional[core.DistributedContext] = None,
+    checkpoint_storage: Optional[Union[str, Dict[str, Any]]] = None,
+    preempt_mode: core.PreemptMode = core.PreemptMode.WorkersAskChief,
+    tensorboard_mode: core.TensorboardMode = core.TensorboardMode.AUTO,
+    # resume: bool = True  # TODO(ilia): optionally control resume behaviour.
+) -> None:
+    global _context
+    global _client
```

Comment:
Maybe this is me being picky, but is it possible to write the code in such a way that explicit context managers are supported?

I think supporting the singleton pattern is fine, but I'd like to also support a pattern that doesn't involve global state.  For example, `det.experimental.client` assumes singleton works for most all cases but `det.experimental.client.Determined()` can be used explicitly if the singleton pattern isn't right for you.

I'm thinking cases like tests would be prime candidates for the explicit context manager case, plus some small subset of users would likely prefer explicit to magical anyway.

My interpretation of this code is that today, core api v2 is definitely not supporting non-singleton patterns, based on these globals, and other things.  But I could be off base, so let me know if I am.